### PR TITLE
Emulator Feature: Bootstrap with default contracts

### PR DESCRIPTION
**Pull request by mwufi**
Closes: FLIP Fest issue 27 https://github.com/onflow/flip-fest/issues/27

## Description

The emulator should be able to deploy contracts when it starts up. We add the following ones:
- NFT Marketplace
- NFT
- FUSD

To get this to work, we do two things:
1. make the emulator aware of the service account's private key -- this allows us to sign transactions!
2. Add the default contracts

Notes:
  - For NonFungibleToken, we depend on the `flow-nft` repo
  - For NFTMarketplace, we add the contract directly -- In the future, we can restructure the `nft-marketplace` repo so that it's easier to pull from

We also include defaults for sig&hash algo (in `start.go`). (Does this address https://github.com/onflow/flow-emulator/issues/85 ?)
```
ServiceKeySigAlgo       string        `default:"ECDSA_P256" flag:"service-sig-algo" info:"service account key signature algorithm"`
ServiceKeyHashAlgo      string        `default:"SHA3_256" flag:"service-hash-algo" info:"service account key hash algorithm"`
```

## Tests
`GO111MODULE=on go run ./cmd/emulator --contracts`
image

`GO111MODULE=on go run ./cmd/emulator --contracts --service-priv-key 92ecc6ca2a30b1da782dc6ca6a6da73ad3d9214adb041b5185ca5196bd46e3c7`
If you run with a custom private key, it reflects that:
image

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by mwufi** [comment]
hello @janezpodhostnik @psiemens @sideninja @turbolent :) tagging just in case you didn't see!

File: cmd/emulator/start/start.go
Comment (janezpodhostnik) [inline cmd/emulator/start/start.go:63]: This can be up for discussion, but I feel the default should be true.

My reasoning is that this will require no change from anyone that is already using the emulator and is depending on the emulator being in a certain state when it starts.

File: server/contracts.go
Comment (janezpodhostnik) [inline server/contracts.go:14]: I would try not to hardcode this. Can you try using this instead:  https://github.com/onflow/flow-go/blob/3932592e17a289880b693e4e57b9044d09dc194a/fvm/bootstrap.go#L927

File: server/contracts/NFTStorefront.cdc
Comment (janezpodhostnik) [inline server/contracts/NFTStorefront.cdc:1]: It would be cool if we could also import this one through a go module like we do the `NonFungibleToken` one 🤔.

That would also mean that the emulator doesn't need the import address replacement logic.

**Review by janezpodhostnik — COMMENTED** [review]
Looks very good overall. Thank you for this!

I left some comments.

File: cmd/emulator/start/start.go
Comment (mwufi) [reply cmd/emulator/start/start.go:63]: @srinjoyc what do you think? Should we make it an "--include-nft-contracts" flag?

File: server/contracts.go
Comment (mwufi) [reply server/contracts.go:14]: Ah! Yes, good idea. Will change to be the fvm address

File: server/contracts/NFTStorefront.cdc
Comment (mwufi) [reply server/contracts/NFTStorefront.cdc:1]: Yeah, it would!! But maybe later? It'd just require making a few changes to the `nft-marketplace` repo, to make it friendlier to use as a module. It'd be really neat to make some kind of interface for structuring new contract repos so that they can be imported easily!! (Maybe like how fvm imports the FungibleToken and other contracts!)

On the plus side, by including this file it shows how easy it is to add your own contracts ;)

File: cmd/emulator/start/start.go
Comment (sideninja) [reply cmd/emulator/start/start.go:63]: I second @janezpodhostnik comment. Although I believe it could be deployed by default we need to be mindful of other devs depending on the current emulator state.

File: blockchain.go
Comment (sideninja) [inline blockchain.go:197]: I have doubts if this API should be removed. I think it would be better if `WithServicePrivateKey` would be added, and that one shouldn't have `servicePublicKey` property as it can be inferred from the private key. This is especially true or cases where you don't need these contracts so the private key is not necessary.

File: blockchain.go
Comment (sideninja) [inline blockchain.go:205]: Passing both public and private keys should be avoided, public keys should be inferred from the private keys.

File: server/contracts.go
Comment (sideninja) [inline server/contracts.go:38]: Prefer using `fmt.Sprintf` instead of string concatenation with `+` operator.

File: server/contracts.go
Comment (sideninja) [inline server/contracts.go:35]: I believe it would be better to use filepath package join function https://pkg.go.dev/path/filepath#Join

File: server/server.go
Comment (sideninja) [inline server/server.go:235]: what if they only pass public key and not private key but use the option to deploy advanced contracts. This edge case should be handled.

File: server/server.go
Comment (sideninja) [inline server/server.go:139]: Use `fmt.Sprintf` instead.

File: server/contracts.go
Comment (sideninja) [inline server/contracts.go:23]: This should be explicitly ignored if not used.

**Comment by sideninja** [comment]
@mwufi Great effort so far, left some comments. Another general comment is don't forget to update docs with any new flags you add.

File: blockchain.go
Comment (mwufi) [reply blockchain.go]: great. I put this back. it turns out that i don't need to do this -- in fact, in regular usage, this function shouldn't be used at all (since we do want to keep the default service private key :)

Comment (sideninja) [reply blockchain.go]: Exactly, that was worrisome to me.

File: server/server.go
Comment (sideninja) [inline server/server.go:233]: Can you help me understand why do you need to change any functionality regarding passing the private keys in order to finish this task? You deploy contracts by using create account functionality which is working with the current master implementation. I'm not saying you shouldn't improve something if you know how just trying to understand better why you want to change this?

**Comment by sideninja** [comment]
Left a question. Also please make sure the CI tests are passing. By taking a quick look it might be you need to tidy the dependencies.

**Comment by mwufi** [comment]
@sideninja i `go mod tidy`'d the merge conflicts! good to go!

**Comment by sideninja** [comment]
> Hey @psiemens can you quickly review [onflow/fusd#2](https://github.com/onflow/fusd/pull/2)? Made a tiny change to merge before adding it to this one!

I will take a look at this @mwufi

**Comment by mwufi** [comment]
Awesome, thanks @sideninja !!! (other team members, feel free to merge, since i don't have merge permissions!)

File: server/contracts.go
Comment (mwufi) [reply server/contracts.go:58]: ah, interesting!! How do you deploy all at once?

Haha I thought it was all right to have different accounts for each one, but they could also go ito the same account... the bigger thing will probably be.. how do you control the accounts once you've deployed the contracts?
- actually it might be a good idea to make them all deploy from the service acct, since that's what you have the keys for, right?

Comment (sideninja) [reply server/contracts.go:58]: Yes, I think that it would be better to deploy on the service account not just because of the keys but also there are for sure a lot of tests/projects, unfortunately, depending on the sequence of addresses used for testing so if you will "use" those addresses to do these deployments we might break something for them. You can deploy them using a single transaction as contracts can be passed as an array, see here: https://pkg.go.dev/github.com/onflow/flow-go-sdk/templates#CreateAccount

Comment (mwufi) [reply server/contracts.go]: done!

image

**Review by sideninja — APPROVED** [review]
Great improvement. I think this is it 🎉

**Comment by mwufi** [comment]
wait, actually i made a mistake!! with the second way, it didn't actually submit the transactions, just created them :( so now i'm figuring out how to run the transactions !!

alternatively, would it be a good idea to create one (1) new account for the contracts? instead of deploying to the service acct?

image

**Comment by sideninja** [comment]
> wait, actually i made a mistake!! with the second way, it didn't actually submit the transactions, just created them :( so now i'm figuring out how to run the transactions !!
> 
> alternatively, would it be a good idea to create one (1) new account for the contracts? instead of deploying to the service acct?
> 
> image

I planned to test this tomorrow, so will keep you updated. We would only merge this after the Flip fest so no worries. 
I would advise you to deploy on service account since you have the keys and for the reasons explained before, people assume second address is something they have always got as a second address and now this would shift that.

File: server/contracts.go
Comment (sideninja) [inline server/contracts.go:51]: Use emulator instance `SendTransaction` function here.

Comment (mwufi) [reply server/contracts.go:51]: oh!!! will give that a try -- right now i'm just using the raw blockchain methods :P

image

Comment (sideninja) [reply server/contracts.go:51]: You can just use `CreateAccount()` method on blockchain and avoid handling this yourself.

Comment (sideninja) [reply server/contracts.go:51]: Yeah you are right sorry, no method like that on the blockchain.

Comment (mwufi) [reply server/contracts.go:107]: Ah, okay! I think this'll do it (now it actually checks the service account before printing, to make sure!)

image

**Pull request merged**
